projects/28_gltf/main.cpp

```diff
@@ -0,0 +1,774 @@
+    grfx::Queue*                                              pQueue,
+    grfx::DescriptorPool*                                     pDescriptorPool,
+    std::vector<Object>*                                      objects,
+    const std::unordered_map<const cgltf_primitive*, size_t>& primitiveToIndex,
+    std::vector<Primitive>*                                   pPrimitives,
+    std::vector<Material>*                                    pMaterials)
+{
+    const size_t nodeCount = data->nodes_count;
+    for (size_t i = 0; i < nodeCount; i++) {
+        const auto& node = data->nodes[i];
+        if (node.mesh == nullptr) {
+            continue;
+        }
+
+        Object item;
+        item.model = compute_object_matrix(&node);
+
+        for (size_t j = 0; j < node.mesh->primitives_count; j++) {
+            const size_t primitive_index = primitiveToIndex.at(&node.mesh->primitives[j]);
+            // FIXME: support meshes with no material.
+            // For now, assign the first available material.
+            PPX_ASSERT_MSG(pMaterials->size() != 0, "Doesn't support GLTF files with no materials.");
+            cgltf_material* mtl            = node.mesh->primitives[j].material;
+            const size_t    material_index = mtl == nullptr ? 0 : std::distance(data->materials, mtl);
+
+            PPX_ASSERT_MSG(primitive_index < pPrimitives->size(), "Invalid GLB file. Primitive index out of range.");
+            PPX_ASSERT_MSG(material_index < pMaterials->size(), "Invalid GLB file. Material index out of range.");
+            Primitive* pPrimitive = &(*pPrimitives)[primitive_index];
+            Material*  pMaterial  = &(*pMaterials)[material_index];
+
+            grfx::DescriptorSet* pDescriptorSet = nullptr;
+            PPX_CHECKED_CALL(pQueue->GetDevice()->AllocateDescriptorSet(pDescriptorPool, pMaterial->pSetLayout, &pDescriptorSet));
+            item.renderables.emplace_back(pMaterial, pPrimitive, pDescriptorSet);
+        }
+
+        // Create uniform buffer.
+        grfx::BufferCreateInfo bufferCreateInfo        = {};
+        bufferCreateInfo.size                          = RoundUp(512, PPX_CONSTANT_BUFFER_ALIGNMENT);
+        bufferCreateInfo.usageFlags.bits.uniformBuffer = true;
+        bufferCreateInfo.memoryUsage                   = grfx::MEMORY_USAGE_CPU_TO_GPU;
+        PPX_CHECKED_CALL(pQueue->GetDevice()->CreateBuffer(&bufferCreateInfo, &item.pUniformBuffer));
+
+        objects->emplace_back(std::move(item));
+    }
+}
+
+void ProjApp::Setup()
+{
+    // Cameras
+    {
+        mCamera = PerspCamera(60.0f, GetWindowAspect());
+    }
+
+    // Create descriptor pool large enough for this project
+    {
+        grfx::DescriptorPoolCreateInfo poolCreateInfo = {};
+        poolCreateInfo.uniformBuffer                  = 512;
+        poolCreateInfo.sampledImage                   = 512;
+        poolCreateInfo.sampler                        = 512;
+        poolCreateInfo.combinedImageSampler           = 512;
```

Comment:
Need to find out how to replace that for DX11/DX12.

Reply:
Is that not working on DX12?

Reply:
The combinedImageSampler is marked as unsupported (https://github.com/google/bigwheels/blob/main/src/ppx/grfx/dx12/dx12_descriptor.cpp#L44)
Actually DX11 and VK are the only 2 APIs running this sample 😅

Reply:
Removing it fails in another location I'm fixing right now (compressed textures are created with a size of 1 or 2, while DX12 requires 4 min). Don't know if I can run the DX12 sample with sponza without that line yet.

Reply:
So same issue as VK on DX12: pool runs out of descriptors of this type, but I cannot set a larger size than the default one.
Might need to fix this.

Reply:
No, just this HLSL shader. I'm not even sure why storageImage + combinedImageSampler is needed. Removing one of the other causes a pool exhaustion. Now that it is a blocker I have a reason to look into it..

Reply:
So turns out, It might be driver-related:
 - on linux, cannot run without setting both COMBINED_IMAGE_SAMPLER & STORAGE_IMAGE to 512
 - on windows (cloudtop), can run the VK version without those 2 lines.

Reply:
Found the reason:
 - radv doesn't handle a per-descriptor pool size, but allocates more or less bytes in a single pool:
 https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/amd/vulkan/radv_descriptor_set.c#L908
 So the real issue was that we allocated more than 512 descriptors for samplers/sampled_image. But it was fine because we allocated less than 512 uniform buffers. Until we reached this limit.
 Adding any other type bumps this global limit, so that's why it worked.
 
 Bumped the descriptor limit to 1024 for the 3 types we need.